### PR TITLE
Add fast BRAS implementation to scib-metrics package.

### PR DESCRIPTION
**Pull request by prauten**
Hi there,

I am Pia, the first author of the preprint "[Metrics Matter: Why We Need to Stop Using Silhouette in Single-Cell Benchmarking](https://www.biorxiv.org/content/10.1101/2025.01.21.634098v1.full-text)" mentioned in issue #193. 

First, thank you for developing and maintaining this package. It is a valuable resource for the community and much needed considering ever-increasing data set sizes.

Our preprint highlights limitations of silhouette-based metrics for evaluating batch effect removal, particularly in the presence of prevalent nested batch effects. To address these challenges, we propose a new metric called BRAS (batch removal adapted silhouette), which redefines the silhouette score to be robust against such effects. Integrating BRAS into your package would allow users to explore what our work suggests is a more reliable alternative to current silhouette-based metrics for benchmarking integration methods.

I have prepared a pull request implementing BRAS with minimal, backward-compatible changes. This implementation builds on the existing package structure and functionality and includes appropriate documentation. I would greatly appreciate the opportunity to collaborate on integrating this contribution!

Let me know if you would like any adjustments or further context.

Best,
Pia

### CODE REVIEW

File: src/scib_metrics/utils/_silhouette.py [hunk 1]
```diff
@@ -43,21 +58,34 @@ def _silhouette_reduce(
...
+        # of the remaining distances we normalise and extract the maximum
+        clust_dists = clust_dists.at[intra_index].set(-jnp.inf)
+        clust_dists /= label_freqs
+        inter_clust_dists = clust_dists.max(axis=1)
+    elif between_cluster_distances == "mean_other":
+        clust_dists = clust_dists.at[intra_index].set(jnp.nan)
+        total_other_dists = jnp.nansum(clust_dists, axis=1)
+        total_other_count = jnp.sum(label_freqs) - label_freqs[jax.lax.dynamic_slice(labels, (start,), (D_chunk_len,))]
+        inter_clust_dists = total_other_dists / total_other_count
+    else:
```
Comment (adamgayoso) [inline src/scib_metrics/utils/_silhouette.py:78]: Would be great to check all options explicitly and have the else raise an error

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,16 @@ and this project adheres to [Semantic Versioning][].
 
 ## 0.6.0 (unreleased)
 
+### Added
+
+- Add batch removal adapted silhouette (BRAS) metric (`scib_metrics.metrics.bras`) {pr}`197`, which addresses limitations of silhouette for scoring batch effect removal.
```
Comment (adamgayoso) [inline CHANGELOG.md:15]:
```suggestion
- Add batch removal adapted silhouette (BRAS) metric ({func}`scib_metrics.metrics.bras`) {pr}`197`, which addresses limitations of silhouette for scoring batch effect removal.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,16 @@ and this project adheres to [Semantic Versioning][].
...
+- Add cosine distance implementation required for BRAS.
+
+### Changed
+
+- Changed `scib_metrics.utils.cdist` to support cosine distance.
```
Comment (adamgayoso) [inline CHANGELOG.md:20]:
```suggestion
- Changed {func}`scib_metrics.utils.cdist` to support cosine distance.
```

File: src/scib_metrics/metrics/_silhouette.py [hunk 4]
```diff
@@ -16,23 +26,35 @@ def silhouette_label(X: np.ndarray, labels: np.ndarray, rescale: bool = True, ch
...
+    rescale: bool = True,
+    chunk_size: int = 256,
+    metric: Literal["euclidean", "cosine"] = "euclidean",
+    between_cluster_distances: Literal["nearest", "mean_other", "furthest"] = "nearest",
 ) -> float:
     """Average silhouette width (ASW) with respect to batch ids within each label :cite:p:`luecken2022benchmarking`.
 
+    Default parameters ('euclidean', 'nearest') match scIB implementation.
+
+    Additional options enable BRAS compatible usage (see `bras()` documentation).
```
Comment (adamgayoso) [inline src/scib_metrics/metrics/_silhouette.py:56]:
```suggestion
    Additional options enable BRAS compatible usage (see :func:`~scib_metrics.metrics.bras` documentation).
```

File: src/scib_metrics/utils/_silhouette.py [hunk 5]
```diff
@@ -9,13 +10,21 @@
...
+    label_freqs: jnp.ndarray,
+    between_cluster_distances: Literal["nearest", "mean_other", "furthest"] = "nearest",
 ) -> tuple[jnp.ndarray, jnp.ndarray]:
     """Accumulate silhouette statistics for vertical chunk of X.
 
-    Follows scikit-learn implementation.
+    Follows scikit-learn implementation with default parameter usage ('nearest').
+
+    Additional options enable BRAS compatible usage, addressing specific limitations of using silhouette in the context
+     of evaluating data integration (see `bras()` documentation).
```
Comment (adamgayoso) [inline src/scib_metrics/utils/_silhouette.py:26]:
```suggestion
     of evaluating data integration (see :func:`~scib_metrics.metrics.bras` documentation).
```

**Review by adamgayoso — CHANGES_REQUESTED** [review]
Thanks! Some small changes suggested. In a follow-up PR, would you like to add this to the benchmarker? Specifically just the dataclass and the enum should be necessary.

**Comment by prauten** [comment]
Hi Adam,

Thanks for your swift response and code reviews! 

It is a great idea to add the BRAS metric to the benchmarker. Once I get to it, I'll open a separate pull request.

Best,
Pia

File: src/scib_metrics/utils/_silhouette.py (hunk 1 above)
Comment (prauten) [reply src/scib_metrics/utils/_silhouette.py:78]: Thanks for the suggestion, I will implement it and update the pull request.

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,6 +16,16 @@ and this project adheres to [Semantic Versioning][].
 
 - Apply default values for benchmarker metrics {pr}`203`.
 
+### Added
```
Comment (adamgayoso) [inline CHANGELOG.md:19]: Can you move these into the unreleased section above?

**Review by adamgayoso — APPROVED** [review]
LGTM. Please move the release notes up for now. Thanks.

File: CHANGELOG.md (hunk 6 above)
Comment (prauten) [reply CHANGELOG.md:13]: I did! Apologies for the oversight. I thought version 0.5.4 had not been released yet, as it doesn't show as "latest" on the main page of the package on GitHub.
